core/blockchain.go

```diff
@@ -2808,41 +2804,82 @@ func (bc *BlockChain) UpdateStakingMetaData(tx *staking.StakingTransaction, root
 }
 
 // ReadBlockRewardAccumulator must only be called on beaconchain
-func (bc *BlockChain) ReadBlockRewardAccumulator(number uint64) (*big.Int, error) {
+func (bc *BlockChain) ReadBlockRewardAccumulator(number uint64) (*votepower.RewardAccumulation, error) {
 	if !bc.chainConfig.IsStaking(shard.Schedule.CalcEpochNumber(number)) {
-		return big.NewInt(0), nil
+		return &votepower.EmptyReward, nil
 	}
 	if cached, ok := bc.blockAccumulatorCache.Get(number); ok {
-		return cached.(*big.Int), nil
+		return cached.(*votepower.RewardAccumulation), nil
 	}
 	return rawdb.ReadBlockRewardAccumulator(bc.db, number)
 }
 
 // WriteBlockRewardAccumulator directly writes the BlockRewardAccumulator value
 // Note: this should only be called once during staking launch.
-func (bc *BlockChain) WriteBlockRewardAccumulator(reward *big.Int, number uint64) error {
-	err := rawdb.WriteBlockRewardAccumulator(bc.db, reward, number)
-	if err != nil {
+func (bc *BlockChain) WriteBlockRewardAccumulator(rewarded *economics.Produced) error {
+	if err := rawdb.WriteBlockRewardAccumulator(bc.db, rewarded); err != nil {
 		return err
 	}
-	bc.blockAccumulatorCache.Add(number, reward)
+	bc.blockAccumulatorCache.Add(rewarded.ReadBlockNumber(), rewarded.Read())
 	return nil
 }
 
 //UpdateBlockRewardAccumulator ..
 // Note: this should only be called within the blockchain insertBlock process.
-func (bc *BlockChain) UpdateBlockRewardAccumulator(diff *big.Int, number uint64) error {
-	current, err := bc.ReadBlockRewardAccumulator(number - 1)
+func (bc *BlockChain) UpdateBlockRewardAccumulator(rewarded *economics.Produced) error {
+	current, err := bc.ReadBlockRewardAccumulator(rewarded.ReadBlockNumber() - 1)
+	newWinners := []votepower.VoterReward{}
 	if err != nil {
 		// one-off fix for pangaea, return after pangaea enter staking.
-		current = big.NewInt(0)
-		bc.WriteBlockRewardAccumulator(current, number)
+		bc.WriteBlockRewardAccumulator(economics.NewNoReward(rewarded.ReadBlockNumber()))
+	}
+	rewardsSoFar := current.ValidatorReward
+	soFarCount := len(rewardsSoFar)
+
+	// Sort by address, then can do binary search
+	sort.SliceStable(rewardsSoFar, func(i, j int) bool {
+		return bytes.Compare(
+			rewardsSoFar[i].Validator.Bytes(),
+			rewardsSoFar[j].Validator.Bytes(),
+		) == -1
+	})
+
+	newlyRewarded, existingTotal := rewarded.ReadRewarded(), rewarded.ReadTotalPayout()
+	existingTotal.Add(existingTotal, rewarded.ReadTotalPayout())
+
+	for i := range newlyRewarded {
```

Comment:
I know you have a tweet on this. But seems it's not the best way, why don't we just key by the validator address+epoch and value with the accumulated rewards? Then there is no need to search. Just O(1) lookup. This search will get very heavy when number of validators get big. and it's happening for every block.